### PR TITLE
feat(gstd): add `builder` module

### LINKED ISSUE #1829: Redesign `gstd`
### Problem to Solve

`gstd` seems to need redesign with functions like `msg::send_delayed_from_reservation_and_wait_for_reply`

### Possible Solution

.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
### Description

Resolves #1829

TL;DR:
```rust
use gstd::{builder, msg};

// before
msg::send_bytes_with_gas_delayed(ActorId::zero(), b"PING", 1_000_000, 0, 60)
    .expect("failed to send msg");

// after
builder::bytes::<_>()
    .to(ActorId::zero())
    .payload_bytes(b"PING")
    .value(0)
    .with_gas(1_000_000)
    .delayed(60)
    .send()
    .expect("failed to send msg");

// before
ProgramGenerator::create_program_with_gas(CodeId::new([0xff; 32]), b"payload goes here", 1_000_000, 0)
    .expect("failed to create program");

// after
builder::bytes::<_>()
    .code_id(CodeId::new([0xff; 32]))
    .payload_bytes(b"payload goes here")
    .value(0)
    .with_gas(1_000_000)
    .create_program()
    .expect("failed to create program");

// before
let future: CreateProgramFuture = ProgramGenerator::create_program_for_reply(
    CodeId::new([0xff; 32]),
    b"wasm code goes here",
    0,
)
.expect("failed to create program");

// after
let future: CreateProgramFuture = builder::bytes::<_>()
    .code_id(CodeId::new([0xff; 32]))
    .payload_bytes(b"wasm code goes here")
    .for_reply() //also available .for_reply_as::<D>
    .create_program()
    .expect("failed to create program");
```

### Advantages
- if you look at the code without an IDE, it's much easier to understand
- easier to add delay, gas limit, reservation to an existing call

### Disadvantages
- complex implementation
- not all combinations of chained calls currently exist, which can lead to runtime errors (we can't verify it at compile time w/o extra `const fn` call)

### TODO
- [ ] delete `unreachable!()` and replace it with `return Err(ContractError::BuilderUsage("msg"))`
- [ ] see how clear the error messages are
- [ ] test some part of the chained calls:
  - should be compiled to zero cost
  - have the same behavior as the original
- [ ] add example programs with new API

@gear-tech/dev 
@gear-tech/ecosystem-devs

### CODE REVIEW

File: gstd/src/builder.rs [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+            value: None,
+            delay: None,
+            gas_limit: None,
+            reservation_id: None,
+        }
+    }
+
+    /// Sets `code_id` which can be used to call
+    /// [`create_program`](Self::create_program) later.
+    pub const fn code_id(mut self, code_id: CodeId) -> Self {
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply gstd/src/builder.rs:161]: I think it would be redundant because we have different setters like
- `pub const fn payload_encode(mut self, payload: &'a Encodable) -> Self` (by ref)
- `pub const fn payload_input(mut self, payload: Range) -> Self` (by value)

Also note that they are marked `const fn` to achieve zero-cost abstraction. Implementation takes much more lines of code than these setters

**Comment by DennisInSky** [comment]
If we've decided to use the builder pattern for making calls, I do not think we have to build it around composing a message. My understanding is that a user wants to make a call, thus they want to build this call, and not a message which will be sent to some extrinsic. The latter is lower level compared to the API we expose now, and to my point of view it will overwhelm the user. The main point of any public API is providing a functionality in the most convinient way for the user. Personally, I would not think of the suggested approach as of an example of the latter. It makes sense to use the builder pattern, but only for the parameters which can be considered optional in terms of that, if the user does not specify a value for them, there is always some default value. For some cases, it means there will be no anything to configure at all.
As a user, in most cases I would like to **call** functionality for sending a message, i.e. it might look like
```rust
builder::send_message(signer, tgt_pid, payload).execute();
```
the above would send a message using max gas limit, 0 value
but if I want to do more customisations, then I could write
```rust
builder::send_message(signer, tgt_pid, payload).with_gas_limit(gas_limit).with_value(42).execute();
```
Hope it is clear what I tried to convey

**Comment by shamilsan** [comment]
> // after
> builder::bytes::<_>()
>     .to(ActorId::zero())
>     .payload_bytes(b"PING")
>     .value(0)
>     .with_gas(1_000_000)
>     .delayed(60)
>     .send()
>     .expect("failed to send msg");

Could we eliminate these `bytes::<_>()` boilerplate calls? Maybe add some shortcut as `msg::builder().to(ActorId::zero()).etc(...)`?

**Comment by shamilsan** [comment]
> It makes sense to use the builder pattern, but only for the parameters which can be considered optional in terms of that, if the user does not specify a value for them, there is always some default value. 

Good point, agree.

**Comment by NikVolf** [comment]
While it is good work, it should not be in `gstd`, it's too opinionated

Would welcome to put it in separate library in `gear-tech` repo

**Comment by shamilsan** [comment]
We've agreed to the following:

- Create a dedicated crate for it (need to choose a proper name).
- Move on to the typestate pattern with a limited set of types and fixed builder methods sequence.

**Comment by breathx** [comment]
Stale PR?

**Comment by StackOverflowExcept1on** [comment]
@breathx Shamil said we shouldn't rush this PR. For example, we have #2705 and #2862. They are aimed at extending functionality of program and msg APIs. So we need to figure out how to make this builder extensible.

**Comment by breathx** [comment]
> @breathx Shamil said we shouldn't rush this PR. For example, we have #2705 and #2862. They are aimed at extending functionality of program and msg APIs. So we need to figure out how to make this builder extensible.

yeah but we may introduce some first impl in different crate inside the workspace

**Comment by breathx** [comment]
To be reopened once finished in this or other repo

**Comment by StackOverflowExcept1on** [comment]
development moved to this repository: https://github.com/gear-foundation/gstd-fluent
